components/datetime/src/format/time_zone.rs

```diff
@@ -37,11 +37,79 @@ where
     T: TimeZoneInput,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write_pattern(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
+        write_zone(self.time_zone_format, self.time_zone, f).map_err(|_| core::fmt::Error)
     }
 }
 
-// TODO(#622) Make this public once TimeZoneFormat is public.
+pub(crate) fn write_zone<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match (&time_zone_format.patterns, &time_zone_format.config) {
+        (Some(_), None) => write_pattern(time_zone_format, time_zone, w),
+        (None, Some(_)) => write_config(time_zone_format, time_zone, w),
+        _ => todo!("Invalid state (both populated, or none populated, shouldn't be possible)"),
+    }
+}
+
+pub(crate) fn write_config<T, W>(
+    time_zone_format: &TimeZoneFormat,
+    time_zone: &T,
+    w: &mut W,
+) -> Result<(), Error>
+where
+    T: TimeZoneInput,
+    W: fmt::Write + ?Sized,
+{
+    match time_zone_format.config.as_ref().unwrap() {
```

Comment:
> suggestion

Now that `TimeZoneFormatConfig` is a `Copy` type, we can remove the `as_ref()` here, because the ownership won't be moved anymore.

Reply:
> suggestion

Instead of using `unwrap()` here, let's use `expect()` which allows you to also provide a message if this fails. 

We could say something like 
```rust
.expect("TimeZoneFormat::write_config() expected Some(config) but got None")
```